dali/operators/generic/erase/erase.cu

```diff
@@ -0,0 +1,162 @@
+                               uint64_t, int64_t, float)
+namespace dali {
+
+namespace detail {
+
+template <int ndim>
+ivec<ndim> to_ivec(const TensorShape<ndim> &shape) {
+  ivec<ndim> result;
+  for (int d = 0; d < ndim; d++) {
+    result[d] = shape[d];
+  }
+  return result;
+}
+
+template <int ndim>
+kernels::ibox<ndim> make_box(const TensorShape<ndim> &lcorner, const TensorShape<ndim> &shape) {
+  auto lc = to_ivec(lcorner);
+  return Box<ndim, int>(lc, lc + to_ivec(shape));
+}
+
+}  // namespace detail
+
+template <typename T, int Dims, int channel_dim>
+class EraseImplGpu : public OpImplBase<GPUBackend> {
+ public:
+  using EraseKernel = kernels::EraseGpu<T, Dims, channel_dim>;
+
+  explicit EraseImplGpu(const OpSpec &spec)
+      : spec_(spec), batch_size_(spec.GetArgument<int>("batch_size")) {
+        kmgr_.Resize<EraseKernel>(1, 1);
+      }
+
+  bool SetupImpl(std::vector<OutputDesc> &output_desc,
+                 const workspace_t<GPUBackend> &ws) override {
+    const auto &input = ws.template InputRef<GPUBackend>(0);
+    auto layout = input.GetLayout();
+    auto type = input.type();
+    auto shape = input.shape();
+    AcquireArgs(ws, shape, layout);
+
+    auto in_view = view<const T, Dims>(input);
+    kmgr_.Initialize<EraseKernel>();
+    ctx_.gpu.stream = ws.stream();
+    kmgr_.Setup<EraseKernel>(0, ctx_, in_view, regions_, make_cspan(fill_values_));
+
+    output_desc.resize(1);
+    output_desc[0] = {shape, input.type()};
+    return true;
+  };
+
+  void RunImpl(workspace_t<GPUBackend> &ws) override {
+    auto input = view<const T, Dims>(ws.template InputRef<GPUBackend>(0));
+    auto output = view<T, Dims>(ws.template OutputRef<GPUBackend>(0));
+    kmgr_.Run<EraseKernel>(0, 0, ctx_, output, input, regions_, make_cspan(fill_values_));
+  }
+
+  void AcquireArgs(const ArgumentWorkspace &ws, TensorListShape<> in_shape,
+                   TensorLayout in_layout) {
+    fill_values_ = spec_.template GetRepeatedArgument<float>("fill_value");
+    DALI_ENFORCE(channel_dim >= 0 || fill_values_.size() <= 1,
+      "If a multi channel fill value is provided, the input layout must have a 'C' dimension");
+    regions_.resize(batch_size_);
+    auto args = detail::GetEraseArgs<T, Dims>(spec_, ws, in_shape, in_layout);
+    for (int i = 0; i < batch_size_; ++i) {
+      regions_[i].reserve(args[i].rois.size());
+      const auto &rois = args[i].rois;
+      for (const auto &roi : args[i].rois) {
+        regions_[i].push_back(detail::make_box(roi.anchor, roi.shape));
+      }
+    }
+  }
+
+ private:
+  OpSpec spec_;
+  int batch_size_ = 0;
+  std::vector<int> axes_;
+  std::vector<kernels::EraseArgs<T, Dims>> args_;
+  std::vector<std::vector<kernels::ibox<Dims>>> regions_;
+  SmallVector<T, 3> fill_values_;
+  kernels::KernelManager kmgr_;
+  kernels::KernelContext ctx_;
+};
+
+namespace detail {
+
+template <typename T, int Dims, int... ChDims>
+struct select_impl {};
```

Comment:
Why does it need to be variadic at all? Is it even correct?
I think the code below is much easier to follow:
```
template <typename T, int Dim, typename... Args>
std::unique_ptr<OpImplBase<GPUBackend>> select_impl(int channel_dim, std::integral_constant<int, -1> chdim, const Args&... args) {
  if (channel_dim == chdim)
    return std::make_unique<EraseImplGPU<T, Dims, ChDim>(args...);
}

template <typename T, int Dim, int ChDim = Dim-1, typename... Args>
std::unique_ptr<OpImplBase<GPUBackend>> select_impl(int channel_dim, std::integral_constant<int, ChDim> chdim, const Args&... args) {
  if (channel_dim == chdim)
    return std::make_unique<EraseImplGPU<T, Dims, ChDim>(args...);
  else
     return select_impl<T, Dim>(channel_dim, std::integral_constant<int, ChDim-1>(), args...);
}
```

Reply:
I limit possible layouts to channel-first, channel-last and no channel, because compilation gets crazy long with all the combinations.